### PR TITLE
[Serialization] Soft-reject swiftmodules built against a different SDK on tagged compilers

### PR DESCRIPTION
**Pull request by xymus**
Rework how the compiler handles swiftmodules built against a different SDK. Instead of raising an error this makes the compiler soft-reject the swiftmodule and silently rebuild from the swiftinterface (as it does with outdated modules). We still get the hard errors for modules with no swiftinterfaces as we can't rebuild them.

Also limit this check to tagged compilers only. This allows dev compilers to be more versatile for testing and production compilers to be more reliable once deployed.

rdar://93257769

---

Note that in practice, this feature is still turned off by default on the driver side.

### CODE REVIEW

**Comment by xymus** [comment]
@swift-ci Please smoke test

File: lib/Serialization/ModuleFileSharedCore.cpp [hunk 1]
```diff
@@ -307,6 +308,30 @@ static ValidationInfo validateControlBlock(
       break;
     case control_block::SDK_NAME: {
       result.sdkName = blobData;
+
+      // Enable this check for tagged compiler or when the
+      // env var is set (for testing).
+      static const char* forceDebugPreSDKRestriction =
+        ::getenv("SWIFT_DEBUG_FORCE_SWIFTMODULE_PER_SDK");
+      if (version::Version::getCurrentCompilerVersion().empty() &&
```
Comment (tshortli) [inline lib/Serialization/ModuleFileSharedCore.cpp:316]: It required a little research for me to understand what this check is doing. I would find it helpful for this predicate to be wrapped in an API that returns a `bool` and documents what it means for a compiler to be tagged or not. I also see a couple other places where such a convenience could be used.

**Review by tshortli — APPROVED** [review]

File: lib/Serialization/ModuleFileSharedCore.cpp (hunk 1 above)
Comment (xymus) [reply lib/Serialization/ModuleFileSharedCore.cpp]: Good idea, the link between the different version services was unclear too. I added the service in the latest commit.

**Comment by xymus** [comment]
@swift-ci Please smoke test

https://github.com/apple/llvm-project/pull/4692

**Comment by xymus** [comment]
@swift-ci Please smoke test

https://github.com/apple/llvm-project/pull/4692

**Pull request merged**